File: backend/nestjs-mongoose/src/modules/auth/auth.service.ts
import { JwtService } from '@nestjs/jwt';
import { Injectable, UnauthorizedException } from '@nestjs/common';

import { iSignin } from './auth.interface';
import { UsersService } from '../users/users.service';
import { SigninUserDto } from 'src/utils/dto/signinUser.dto';
import { CreateUserDto } from 'src/utils/dto/createUser.dto';
import { comparePassword, hashPassword } from 'src/utils/helper';

@Injectable()
export class AuthService {
  constructor(
    private jwtService: JwtService,
    private usersService: UsersService,
  ) {}

  async signUp(payload: CreateUserDto): Promise<iSignin> {
    const query = { email: payload.email };
    const user = await this.usersService.findUser({ query });

    if (user) {
      throw new UnauthorizedException('Email already exists');
    }

    const hashedPassword = await hashPassword(payload.password);
    payload.password = hashedPassword;

    const createdUser = await this.usersService.createUser(payload);

    const result = { name: createdUser.name, email: createdUser.email };

    return {
      ...result,
      cookie: await this.generateCookieToken({
        ...result,
        _id: createdUser._id,
      }),
    };
  }

  async signIn(payload: SigninUserDto): Promise<iSignin> {
    const query = { email: payload.email };
    const user = await this.usersService.findUser({
      query,
      select: 'name email password',
    });

    if (!user) {
      throw new UnauthorizedException('Invalid email or password');
    }

    const passwordMatch = await comparePassword(payload.password, user.password);

    if (!passwordMatch) {
      throw new UnauthorizedException('Invalid email or password');
    }

    const result = { name: user.name, email: user.email };

    return {
      ...result,
      cookie: await this.generateCookieToken({ ...result, _id: user._id }),
    };
  }

  async generateCookieToken(payload: any): Promise<string> {
    return this.jwtService.signAsync(payload);
  }
}

File: backend/python-fastapi/app/modules/auth/auth_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException, status

from ..users.user_service import UsersService
from .auth_dto import CreateUserDTO, SigninUserDTO
from app.utils.helper import hash_password, create_jwt_access_token, verify_password


class AuthService:

    async def sign_up(db: Session, payload: CreateUserDTO):
        query = {"email": payload.email}
        user = UsersService.find_user(db, query)

        if user:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Email already exists",
            )

        hashedPassword = hash_password(payload.password)
        payload.password = hashedPassword

        createdUser = UsersService.create_user(db, payload)

        result = {"name": createdUser.name, "email": createdUser.email}

        return {
            **result,
            "cookie": create_jwt_access_token(
                {
                    **result,
                    "id": createdUser.id,
                }
            ),
        }

    async def sign_in(db: Session, payload: SigninUserDTO):
        query = {"email": payload.email}
        user = UsersService.find_user(
            db, query, select=["id", "name", "email", "password"]
        )

        if not user:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Incorrect email or password",
            )

        password_Match = verify_password(payload.password, user.password)

        if not password_Match:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Incorrect email or password",
            )

        result = {"name": user.name, "email": user.email}

        return {
            **result,
            "cookie": create_jwt_access_token({**result, "id": user.id}),
        }
